File: sce-rust-runtime/src/scripting/i_script_engine.rs
// SPDX-License-Identifier: LGPL-2.1-or-later WITH LicenseRef-SCE-Linking-Exception OR LicenseRef-SCE-Commercial
// SPDX-FileCopyrightText: Copyright (c) 2025 newmassrael

//! [`IScriptEngine`] — the script engine trait, 1:1 port of C++ `IScriptEngine.h`.
//!
//! Port rules:
//! - C++ `std::future<ScriptResult>` return types become synchronous `Result<ScriptValue, ScriptError>`.
//!   C++ resolves futures via `.get()` in practice (see callers in
//!   `sce/common/AssignmentExecutionHelper.cpp` etc.), so the sync shape preserves semantics.
//! - C++ `const std::string&` parameters become `&str`.
//! - C++ `const std::vector<...>&` parameters become `&[...]`.
//! - C++ inheritance `ISessionLifecycle` is flattened into this trait.
//! - C++ `std::function` callbacks become `Box<dyn Fn(...) + Send + Sync>`.

use std::collections::HashMap;

use thiserror::Error;

/// A value that can cross the script engine boundary.
///
/// Matches C++ `SCE::ScriptValue` (a `std::variant<...>`-like discriminated union).
/// Generated code converts datamodel variable values to/from `ScriptValue` at assignment
/// and guard evaluation sites.
#[derive(Debug, Clone, PartialEq)]
pub enum ScriptValue {
    /// JavaScript `null` / Lua `nil`.
    Null,
    /// `undefined` (JavaScript only; Lua treats as `nil`).
    Undefined,
    /// Boolean.
    Bool(bool),
    /// 64-bit signed integer.
    Int(i64),
    /// Double-precision float.
    Double(f64),
    /// UTF-8 string.
    String(String),
    /// Ordered array of values.
    Array(Vec<ScriptValue>),
    /// Key-value object (iteration order is implementation-defined).
    Object(HashMap<String, ScriptValue>),
    /// Opaque DOM reference (W3C SCXML B.2 — XML datamodel support).
    /// Stored as the original XML string; the script engine parses on demand.
    Dom(String),
}

impl ScriptValue {
    /// Attempt to coerce to `bool` per SCXML truthiness rules
    /// (W3C SCXML B.2.3: ECMAScript truthy/falsy semantics).
    pub fn to_bool(&self) -> bool {
        match self {
            ScriptValue::Null | ScriptValue::Undefined => false,
            ScriptValue::Bool(b) => *b,
            ScriptValue::Int(i) => *i != 0,
            ScriptValue::Double(f) => *f != 0.0 && !f.is_nan(),
            ScriptValue::String(s) => !s.is_empty(),
            ScriptValue::Array(_) | ScriptValue::Object(_) | ScriptValue::Dom(_) => true,
        }
    }

    /// W3C SCXML 5.5: Convert to a Lua literal string representation.
    ///
    /// Used by donedata evaluation to produce event data strings that the Lua
    /// script engine can parse back via `return <literal>`. Ports the C++
    /// `DoneDataHelper::convertScriptValueToJson` but emits Lua syntax instead
    /// of JSON so the Lua engine's `load("return " .. data)` path works.
    pub fn to_lua_literal(&self) -> String {
        match self {
            ScriptValue::Null | ScriptValue::Undefined => "nil".to_string(),
            ScriptValue::Bool(b) => if *b { "true" } else { "false" }.to_string(),
            ScriptValue::Int(i) => i.to_string(),
            ScriptValue::Double(f) => {
                if f.fract() == 0.0 && f.is_finite() {
                    format!("{:.1}", f)
                } else {
                    format!("{}", f)
                }
            }
            ScriptValue::String(s) => {
                let escaped = s
                    .replace('\\', "\\\\")
                    .replace('"', "\\\"")
                    .replace('\n', "\\n")
                    .replace('\r', "\\r")
                    .replace('\t', "\\t");
                format!("\"{}\"", escaped)
            }
            ScriptValue::Array(arr) => {
                let items: Vec<String> = arr.iter().map(|v| v.to_lua_literal()).collect();
                format!("{{{}}}", items.join(", "))
            }
            ScriptValue::Object(map) => {
                let items: Vec<String> = map
                    .iter()
                    .map(|(k, v)| format!("[\"{}\"] = {}", k, v.to_lua_literal()))
                    .collect();
                format!("{{{}}}", items.join(", "))
            }
            ScriptValue::Dom(s) => {
                let escaped = s.replace('\\', "\\\\").replace('"', "\\\"");
                format!("\"{}\"", escaped)
            }
        }
    }
}

/// Script engine error. 1:1 port of C++ `ScriptResult::error` field.
#[derive(Debug, Clone, Error)]
pub enum ScriptError {
    /// Syntax error during compilation (W3C SCXML 5.9: raises `error.execution`).
    #[error("script syntax error: {0}")]
    SyntaxError(String),

    /// Runtime error during evaluation (W3C SCXML 5.9: raises `error.execution`).
    #[error("script runtime error: {0}")]
    RuntimeError(String),

    /// Attempted to access a session that does not exist.
    #[error("session not found: {0}")]
    SessionNotFound(String),

    /// Variable not declared in the session's datamodel.
    #[error("variable not declared: {0}")]
    VariableNotDeclared(String),

    /// Attempted to assign to a read-only system variable (_event, _sessionid, etc.).
    #[error("cannot assign to system variable: {0}")]
    ReadOnlySystemVariable(String),

    /// Engine is not initialized; call [`IScriptEngine::initialize`] first.
    #[error("script engine not initialized")]
    NotInitialized,

    /// Engine-specific error (e.g., mlua error wrapper).
    #[error("engine error: {0}")]
    EngineError(String),
}

/// Convenience type alias.
pub type ScriptResult<T> = Result<T, ScriptError>;

/// A native Rust function callable from script code.
///
/// Matches C++ `NativeMethod = std::function<ScriptValue(const std::vector<ScriptValue>&)>`.
pub type NativeMethod = Box<dyn Fn(&[ScriptValue]) -> ScriptValue + Send + Sync>;

/// Callback for resolving the SCXML `In(stateId)` predicate (W3C SCXML 5.9.2).
///
/// Receives a state ID string and returns `true` if that state is in the engine's
/// current active configuration. Matches the C++ `std::function<bool(const std::string &)>`
/// signature passed to `setStateQueryCallback`.
pub type StateQueryCallback = Box<dyn Fn(&str) -> bool + Send + Sync>;

/// The script engine trait — 1:1 port of C++ `SCE::IScriptEngine`.
///
/// Implementations (`sce-rust-lua`, future `sce-rust-quickjs`) provide ECMAScript
/// evaluation for W3C SCXML B.1 datamodel support. Engine DI Parity RFC
/// (Path B+): consumers pass an `Arc<dyn IScriptEngine>` to the generated
/// `Policy::new(engine)` constructor; there is no process-global singleton.
pub trait IScriptEngine: Send + Sync {
    // ════════════════════════════════════════
    // Core Script Execution
    // ════════════════════════════════════════

    /// Execute a script block in the specified session (W3C SCXML 3.8.6).
    ///
    /// Matches C++ `executeScript(const string&, const string&) -> future<ScriptResult>`.
    /// The future is resolved synchronously — returns the final expression value.
    fn execute_script(&self, session_id: &str, script: &str) -> ScriptResult<ScriptValue>;

    /// Evaluate an expression and return its value (W3C SCXML 5.3).
    ///
    /// Matches C++ `evaluateExpression(...) -> future<ScriptResult>`. Used for
    /// variable init, `<param expr="...">`, guard conditions, etc.
    fn evaluate_expression(&self, session_id: &str, expression: &str) -> ScriptResult<ScriptValue>;

    /// Validate expression syntax without executing.
    ///
    /// Matches C++ `validateExpression(...) -> future<ScriptResult>`. Returns `true`
    /// if the syntax is valid. Used by the datamodel parser for early error detection.
    fn validate_expression(&self, session_id: &str, expression: &str) -> ScriptResult<bool>;

    // ════════════════════════════════════════
    // Variable Management
    // ════════════════════════════════════════

    /// Set a variable in the specified session (W3C SCXML 5.3).
    fn set_variable(&self, session_id: &str, name: &str, value: ScriptValue) -> ScriptResult<()>;

    /// Get a variable from the specified session (W3C SCXML 5.3).
    fn get_variable(&self, session_id: &str, name: &str) -> ScriptResult<ScriptValue>;

    /// Set a variable to an XML DOM object parsed from the given XML content (W3C SCXML B.2).
    ///
    /// Matches C++ `setVariableAsDOM`. Used for `<data src="...xml">` loading and
    /// inline `<content>` with XML payloads.
    fn set_variable_as_dom(
        &self,
        session_id: &str,
        name: &str,
        xml_content: &str,
    ) -> ScriptResult<()>;

    /// Check if a variable is declared in the session scope (W3C SCXML 4.6 / 6.4).
    ///
    /// Returns `true` if declared even if the value is `null`/`undefined`. Used by
    /// foreach to distinguish declared-but-empty from undeclared variables.
    fn has_variable(&self, session_id: &str, name: &str) -> bool;

    /// Check if a variable was pre-initialized (e.g., by invoke `<param>`).
    ///
    /// Matches C++ `isVariablePreInitialized`. Used by datamodel init to skip
    /// re-initializing variables already set by the parent SM.
    fn is_variable_pre_initialized(&self, session_id: &str, name: &str) -> bool;

    // ════════════════════════════════════════
    // SCXML-specific Features
    // ════════════════════════════════════════

    /// Set up SCXML system variables: `_sessionid`, `_name`, `_ioprocessors` (W3C SCXML 5.10).
    fn setup_system_variables(
        &self,
        session_id: &str,
        session_name: &str,
        io_processors: &[String],
    ) -> ScriptResult<()>;

    /// Set the `_event` system variable for the currently-processing event (W3C SCXML 5.10).
    ///
    /// Called before guard evaluation and action execution for each event. The
    /// individual-fields overload matches C++ `setCurrentEvent(sessionId, eventName, ...)`.
    // Justification (clippy::too_many_arguments): 8 fields mirror the C++
    // `IScriptEngine::setCurrentEvent` signature; bundling into a struct would
    // break the documented 1:1 port and require every script engine
    // implementation to redo the field plumbing.
    #[allow(clippy::too_many_arguments)]
    fn set_current_event(
        &self,
        session_id: &str,
        event_name: &str,
        event_data: &str,
        event_type: &str,
        send_id: &str,
        origin: &str,
        origin_type: &str,
        invoke_id: &str,
    ) -> ScriptResult<()>;

    // ════════════════════════════════════════
    // Global Function Management
    // ════════════════════════════════════════

    /// Register a native function accessible from script code.
    ///
    /// Matches C++ `registerGlobalFunction`. Returns `true` on success. Used by
    /// runtime integrations (e.g., logging, state query callbacks) to expose Rust
    /// code to generated Lua/JS expressions.
    fn register_global_function(&self, function_name: &str, callback: NativeMethod) -> bool;

    // ════════════════════════════════════════
    // Native Object Binding
    // ════════════════════════════════════════

    /// Bind a native object with method table as a script-accessible object.
    ///
    /// Matches C++ `bindNativeObject`. Creates `objectName.methodName()` accessor
    /// in the session's datamodel scope. Used for Dependency Injection of native
    /// hardware/service shims into SCXML datamodels.
    fn bind_native_object(
        &self,
        session_id: &str,
        object_name: &str,
        methods: Vec<(String, NativeMethod)>,
    ) -> bool;

    // ════════════════════════════════════════
    // Engine Information
    // ════════════════════════════════════════

    /// Human-readable engine info string (name + version).
    fn get_engine_info(&self) -> String;

    /// Current memory usage in bytes (heap allocated by the script engine).
    fn get_memory_usage(&self) -> usize;

    /// Force garbage collection.
    fn collect_garbage(&self);

    // ════════════════════════════════════════
    // State Query Callback (W3C SCXML 5.9.2 In() predicate)
    // ════════════════════════════════════════

    /// Register a callback that resolves the SCXML `In()` predicate.
    ///
    /// See [`StateQueryCallback`] for the callback signature.
    ///
    /// Matches C++ `setStateQueryCallback`. The callback takes a state ID string
    /// and returns `true` if the state is in the current active configuration.
    /// Passing `None` unregisters the callback.
    fn set_state_query_callback(&self, session_id: &str, callback: Option<StateQueryCallback>);

    // ════════════════════════════════════════
    // Engine Lifecycle
    // ════════════════════════════════════════

    /// Initialize the engine. Called once per process. Returns `true` on success.
    fn initialize(&self) -> bool;

    /// Shutdown and release all sessions. Called during program exit.
    fn shutdown(&self);

    /// Whether the engine has been successfully initialized.
    fn is_initialized(&self) -> bool;

    /// Reset engine state for test isolation.
    ///
    /// Destroys all sessions, clears registered functions/callbacks, then
    /// re-initializes. Matches C++ `reset()`.
    fn reset(&self);

    // ════════════════════════════════════════
    // Session Lifecycle (C++: inherited from ISessionLifecycle)
    // ════════════════════════════════════════

    /// Create a new evaluation session.
    ///
    /// Each state machine instance has its own session with an isolated variable scope.
    fn create_session(&self, session_id: &str);

    /// Destroy a session and release its resources.
    fn destroy_session(&self, session_id: &str);

    /// Check whether a session exists.
    fn has_session(&self, session_id: &str) -> bool;
}

File: backends/python/runtime/sce_runtime/scripting/i_script_engine.py
# SPDX-License-Identifier: LGPL-2.1-or-later WITH LicenseRef-SCE-Linking-Exception OR LicenseRef-SCE-Commercial
# SPDX-FileCopyrightText: Copyright (c) 2026 newmassrael
"""IScriptEngine — script engine contract, 1:1 port of C++ `IScriptEngine.h`.

Mirrors `sce_rust_runtime::IScriptEngine` and the Go / Kotlin / C11 equivalents.
Generated state machine code calls into this trait through the
`ScriptEngineProvider` singleton (no runtime dependency injection); each
`Engine` instance creates / destroys its own session id.

The Lua-backed default implementation lives in `lua_engine.py`. The trait
is intentionally engine-neutral so a future QuickJS-backed impl could
register itself in the same provider slot.
"""

from __future__ import annotations

from abc import ABC, abstractmethod
from dataclasses import dataclass, field
from enum import Enum
from typing import Any, Callable, Dict, List, Optional

from ..io_processors import IoProcessorDescriptor


def _json_string(text: str) -> str:
    """A JSON string literal. Ports C++ `DoneDataHelper::escapeJsonString`.

    Spelled out rather than delegating to `json.dumps` so the escape set
    is the one the other five backends apply — `json.dumps` also escapes
    non-ASCII by default, which would make the same payload differ
    between backends byte for byte.
    """
    escaped = (
        text.replace("\\", "\\\\")
        .replace('"', '\\"')
        .replace("\n", "\\n")
        .replace("\r", "\\r")
        .replace("\t", "\\t")
        .replace("\b", "\\b")
        .replace("\f", "\\f")
    )
    return f'"{escaped}"'


class ScriptValueKind(Enum):
    """Discriminator for `ScriptValue`. Matches the C++ `std::variant`
    case set so DOM payloads can round-trip without losing type."""

    NULL = "null"
    UNDEFINED = "undefined"
    BOOL = "bool"
    INT = "int"
    DOUBLE = "double"
    STRING = "string"
    ARRAY = "array"
    OBJECT = "object"
    DOM = "dom"


@dataclass
class ScriptValue:
    """A value that can cross the script-engine boundary.

    Matches `backends/rust/runtime/src/scripting/i_script_engine.rs::ScriptValue`
    field-for-field. Generated code converts datamodel values to/from
    `ScriptValue` at assignment and guard-evaluation sites. The kind tag
    is explicit so JavaScript's `null` vs `undefined` distinction
    survives across the Lua boundary (Lua collapses both to `nil`)."""

    kind: ScriptValueKind = ScriptValueKind.NULL
    bool_val: bool = False
    int_val: int = 0
    double_val: float = 0.0
    string_val: str = ""
    array_val: List["ScriptValue"] = field(default_factory=list)
    object_val: Dict[str, "ScriptValue"] = field(default_factory=dict)
    dom_val: str = ""

    @staticmethod
    def null() -> "ScriptValue":
        return ScriptValue(kind=ScriptValueKind.NULL)

    @staticmethod
    def undefined() -> "ScriptValue":
        return ScriptValue(kind=ScriptValueKind.UNDEFINED)

    @staticmethod
    def of(value: Any) -> "ScriptValue":
        """W3C SCXML 5.10 — coerce a Python value into a `ScriptValue`
        for the script-engine boundary. `None` always maps to NULL (Lua
        collapses null/undefined to `nil` so the round-trip is lossy at
        that side; callers needing the distinction must construct
        `ScriptValue.undefined()` explicitly)."""
        if value is None:
            return ScriptValue.null()
        if isinstance(value, bool):
            return ScriptValue(kind=ScriptValueKind.BOOL, bool_val=value)
        if isinstance(value, int):
            return ScriptValue(kind=ScriptValueKind.INT, int_val=value)
        if isinstance(value, float):
            return ScriptValue(kind=ScriptValueKind.DOUBLE, double_val=value)
        if isinstance(value, str):
            return ScriptValue(kind=ScriptValueKind.STRING, string_val=value)
        if isinstance(value, list):
            return ScriptValue(
                kind=ScriptValueKind.ARRAY,
                array_val=[ScriptValue.of(v) for v in value],
            )
        if isinstance(value, dict):
            return ScriptValue(
                kind=ScriptValueKind.OBJECT,
                object_val={str(k): ScriptValue.of(v) for k, v in value.items()},
            )
        # Fallback: stringify any other Python type so it round-trips
        # through the Lua engine without losing data.
        return ScriptValue(kind=ScriptValueKind.STRING, string_val=str(value))

    def to_python(self) -> Any:
        """W3C SCXML 5.10 — unwrap to the equivalent Python type for
        consumption by the engine (`_event.data` and friends)."""
        if self.kind is ScriptValueKind.NULL or self.kind is ScriptValueKind.UNDEFINED:
            return None
        if self.kind is ScriptValueKind.BOOL:
            return self.bool_val
        if self.kind is ScriptValueKind.INT:
            return self.int_val
        if self.kind is ScriptValueKind.DOUBLE:
            return self.double_val
        if self.kind is ScriptValueKind.STRING:
            return self.string_val
        if self.kind is ScriptValueKind.ARRAY:
            return [v.to_python() for v in self.array_val]
        if self.kind is ScriptValueKind.OBJECT:
            return {k: v.to_python() for k, v in self.object_val.items()}
        if self.kind is ScriptValueKind.DOM:
            return self.dom_val
        return None

    def to_wire_string(self) -> str:
        """W3C SCXML C.2 — this value as the text a form-encoded param
        carries.

        The BasicHTTP Event I/O Processor sends each `<param>` as one
        `name=value` pair, so the value crosses as *text* and the
        receiving end hands that text to `_event.data`; no script engine
        reads it at either end. That is why this is neither of its two
        neighbours: `to_json_literal` would wrap a string in quotes that
        are not part of it, and an engine literal (this channel's lives
        inside its own engine, `lua_engine._python_to_lua_literal`)
        would put the sender's *language* on the wire.

        The rendering is ECMAScript's `String(value)` (§scxml-B-1 makes
        the data model ECMAScript) with the two amendments C++
        `ScriptResultUtils::resultToString` already made: absence renders
        empty rather than as a word (§scxml-C-1), and a structured value
        renders as JSON, because a receiver that is not a script engine
        has no other reading of it.

        What this replaced was `str(value.to_python())`, which is
        Python's spelling and not the document's — `True` for a boolean
        the other channels spell `true`, `None` for an absence the wire
        reads as empty, `[1, 2]` with a space JSON does not have.
        """
        if self.kind is ScriptValueKind.NULL or self.kind is ScriptValueKind.UNDEFINED:
            return ""
        if self.kind is ScriptValueKind.BOOL:
            return "true" if self.bool_val else "false"
        if self.kind is ScriptValueKind.INT:
            return str(self.int_val)
        if self.kind is ScriptValueKind.DOUBLE:
            value = self.double_val
            if value != value:
                return "NaN"
            if value == float("inf"):
                return "Infinity"
            if value == float("-inf"):
                return "-Infinity"
            if value == int(value) and abs(value) < 1e15:
                # ECMAScript String(5) is "5"; a `.0` tail is Python's
                # spelling of the number, not the document's.
                return str(int(value))
            return repr(value)
        if self.kind is ScriptValueKind.STRING:
            # Already text: quoting it would deliver characters the
            # document never wrote.
            return self.string_val
        if self.kind is ScriptValueKind.DOM:
            # The receiving end parses XML from document text.
            return self.dom_val or ""
        return self.to_json_literal()

    def to_json_literal(self) -> str:
        """Render this value for a wire that leaves the ECMAScript data
        model.

        The clause cited in the body names JSON as that serialisation —
        it is what the BasicHTTP Event I/O Processor sends — and an event
        payload always leaves the data model: the reader is another
        dequeue, often another session, in a mesh another process running
        another backend.

        This is the counterpart of an engine literal, and the difference
        is the point. A literal is *source*, so reading it back requires
        the receiver to RUN it — which made `_event.data` mean one thing
        on a Lua backend and another on a JavaScript one, and made any
        payload executable at the far end. Ports the C++
        `scriptValueToJson` in `sce/src/common/EventDataHelper.cpp`.

        Object keys are sorted so equal content produces equal bytes,
        matching the C++ `std::map` original and the Rust/Go ports.
        """
        # §scxml-B-2-9: a value that has to leave the ECMAScript data model
        # is serialized to JSON, which reconstructs it in full rather than
        # falling back to a lossy platform format.
        if self.kind is ScriptValueKind.NULL or self.kind is ScriptValueKind.UNDEFINED:
            # JSON has no `undefined`; the C++ port maps both to null.
            return "null"
        if self.kind is ScriptValueKind.BOOL:
            return "true" if self.bool_val else "false"
        if self.kind is ScriptValueKind.INT:
            return str(self.int_val)
        if self.kind is ScriptValueKind.DOUBLE:
            value = self.double_val
            if value != value or value in (float("inf"), float("-inf")):
                # RFC 8259 has no spelling for NaN or the infinities.
                return "null"
            if value == int(value) and abs(value) < 1e15:
                return str(int(value))
            return repr(value)
        if self.kind is ScriptValueKind.STRING:
            return _json_string(self.string_val)
        if self.kind is ScriptValueKind.ARRAY:
            return "[" + ",".join(v.to_json_literal() for v in self.array_val) + "]"
        if self.kind is ScriptValueKind.OBJECT:
            return (
                "{"
                + ",".join(
                    f"{_json_string(k)}:{self.object_val[k].to_json_literal()}"
                    for k in sorted(self.object_val)
                )
                + "}"
            )
        if self.kind is ScriptValueKind.DOM:
            # The data model reads XML into a DOM at the *receiving* end,
            # from the document text — so a DOM that reaches here crosses
            # as that text.
            return _json_string(self.dom_val or "")
        return "null"

    def to_bool(self) -> bool:
        """W3C SCXML B.2.3 — ECMAScript truthiness. Falsy: null,
        undefined, false, 0, NaN, empty string. Everything else truthy.
        Used when a guard's `evaluate_expression` returns a non-bool
        type."""
        if self.kind is ScriptValueKind.NULL or self.kind is ScriptValueKind.UNDEFINED:
            return False
        if self.kind is ScriptValueKind.BOOL:
            return self.bool_val
        if self.kind is ScriptValueKind.INT:
            return self.int_val != 0
        if self.kind is ScriptValueKind.DOUBLE:
            return self.double_val != 0.0 and self.double_val == self.double_val
        if self.kind is ScriptValueKind.STRING:
            return bool(self.string_val)
        return True


class ScriptError(Exception):
    """Base class for script-engine failures (W3C SCXML 5.9 maps these
    onto `error.execution`)."""


class ScriptSyntaxError(ScriptError):
    """Expression / script body failed to parse."""


class ScriptRuntimeError(ScriptError):
    """Evaluation raised at runtime (TypeError, ReferenceError, …)."""


class SessionNotFoundError(ScriptError):
    """Operation referenced an unknown session id."""


class VariableNotDeclaredError(ScriptError):
    """W3C SCXML 5.4 — `<assign location>` referenced a name absent from
    the document's static datamodel schema. The generated engine
    translates this into `error.execution`."""


class ReadOnlySystemVariableError(ScriptError):
    """Caller attempted to overwrite `_event` / `_sessionid` / etc."""


NativeMethod = Callable[[List[ScriptValue]], ScriptValue]
StateQueryCallback = Callable[[str], bool]


@dataclass
class SetCurrentEventArgs:
    """Parameter object for the W3C SCXML 5.10 `set_current_event` boundary.

    Bundles the seven `_event.*` metadata fields (name + 6 metadata) that
    every script engine impl must surface before guard evaluation / action
    execution. Cross-language siblings: `SCE::SetCurrentEventArgs` in C++
    and `sce_rust_runtime::SetCurrentEventArgs` in Rust."""

    event_name: str
    event_data: str = ""
    event_type: str = "internal"
    send_id: str = ""
    origin: str = ""
    origin_type: str = ""
    invoke_id: str = ""


class IScriptEngine(ABC):
    """Script engine contract — 1:1 port of C++ `IScriptEngine.h` and
    `sce_rust_runtime::IScriptEngine`. Implementations (`LuaScriptEngine`
    in `lua_engine.py`, future QuickJS-backed) provide ECMAScript
    evaluation for the W3C SCXML B.2 datamodel. Generated `*_sm.py`
    modules invoke this through `ScriptEngineProvider.get()` — there is
    no runtime DI."""

    # ── Core script execution ──────────────────────────────────────

    @abstractmethod
    def execute_script(self, session_id: str, script: str) -> ScriptValue:
        """W3C SCXML 5.5 — run a `<script>` body. Returns the value of
        the last expression (or NULL when the script is statement-only).
        Syntax / runtime errors raise the matching `ScriptError`
        subclass."""

    @abstractmethod
    def evaluate_expression(self, session_id: str, expression: str) -> ScriptValue:
        """W3C SCXML 5.3 — evaluate an expression (cond / `<assign expr>`
        / `<param expr>` / `<send eventexpr>` / …) and return the
        result."""

    @abstractmethod
    def validate_expression(self, session_id: str, expression: str) -> bool:
        """Parse-only check used by the datamodel loader to surface
        syntax errors during `<data>` registration."""

    # ── Variable management ────────────────────────────────────────

    @abstractmethod
    def declare_variable(
        self, session_id: str, name: str, initial_value: ScriptValue
    ) -> None:
        """W3C SCXML 5.3 — register `<data id>` schema and bind the
        initial value. Called once per `<data>` element during
        datamodel init. Subsequent `<assign>` operations target the
        same name via `set_variable`."""

    @abstractmethod
    def set_variable(
        self, session_id: str, name: str, value: ScriptValue
    ) -> None:
        """W3C SCXML 5.3 — assign a `ScriptValue` to a declared
        location. Raises `VariableNotDeclaredError` when the name is
        absent from the session schema (mirrors C++ behaviour)."""

    @abstractmethod
    def get_variable(self, session_id: str, name: str) -> ScriptValue:
        """W3C SCXML 5.3 — read a variable. Returns UNDEFINED when the
        name is declared but unassigned, NULL when the lookup follows a
        dotted path into an absent member."""

    @abstractmethod
    def set_variable_as_dom(
        self, session_id: str, name: str, xml_content: str
    ) -> None:
        """W3C SCXML B.2 — parse an XML document and bind it as a
        DOM-style variable. Used by `<data src="…xml">` and
        `<content>` with XML payloads."""

    @abstractmethod
    def has_variable(self, session_id: str, name: str) -> bool:
        """Variable declared in the session scope (W3C SCXML 4.6 / 6.4
        — `<foreach>` distinguishes declared-but-empty from undeclared)."""

    # ── SCXML-specific features ────────────────────────────────────

    @abstractmethod
    def setup_system_variables(
        self,
        session_id: str,
        session_name: str,
        io_processors: List[IoProcessorDescriptor],
    ) -> None:
        """W3C SCXML 5.10 — bind `_sessionid`, `_name`, `_ioprocessors`
        into the session scope. Called once per session right after
        `create_session`.

        The descriptors arrive fully resolved from
        `sce_runtime.io_processors.build`. An implementation files each one
        under its name with its location and invents neither, so
        `_ioprocessors` reads identically whichever engine backs the
        session."""

    @abstractmethod
    def set_current_event(
        self,
        session_id: str,
        args: "SetCurrentEventArgs",
    ) -> None:
        """W3C SCXML 5.10 — bind the `_event` system variable for the
        currently-processing event. Called by the runtime before guard
        evaluation and action execution for each event. The 7 metadata
        fields are bundled into a [SetCurrentEventArgs] mirroring the
        C++ `SCE::SetCurrentEventArgs` struct and the Rust
        `SetCurrentEventArgs<'a>` parameter object."""

    # ── Global functions / native bindings ─────────────────────────

    @abstractmethod
    def register_global_function(
        self, function_name: str, callback: NativeMethod
    ) -> bool:
        """Expose a native Python callback at script-engine scope.
        Used for `log_hook` integration and similar."""

    @abstractmethod
    def set_state_query_callback(
        self, session_id: str, callback: Optional[StateQueryCallback]
    ) -> None:
        """W3C SCXML 5.9.2 — register the `In(state)` predicate
        resolver for the session. Passing `None` unregisters."""

    # ── Engine + session lifecycle ─────────────────────────────────

    @abstractmethod
    def initialize(self) -> bool:
        """Initialise the script engine. Idempotent; returns `True`
        once the engine is ready."""

    @abstractmethod
    def shutdown(self) -> None:
        """Release all sessions and engine resources."""

    @abstractmethod
    def is_initialized(self) -> bool: ...

    @abstractmethod
    def reset(self) -> None:
        """Destroy all sessions, clear registered callbacks, then
        re-initialise. Used for test isolation."""

    @abstractmethod
    def create_session(self, session_id: str) -> None: ...

    @abstractmethod
    def destroy_session(self, session_id: str) -> None: ...

    @abstractmethod
    def has_session(self, session_id: str) -> bool: ...

    # ── Engine info ────────────────────────────────────────────────

    @abstractmethod
    def get_engine_info(self) -> str: ...
